README.md
# Detecting grasping positions with deep neural networks using RGB images


- Use grasp_detection for training model. This is for test.
grasp_detection Link : https://github.com/Juna2/grasp_detection

- Most of the files in this repo is from his work except test.py

- test.py loads pretrained model in 

    /home/<your_path>/catkin_ws/src/ros_grasp_detection/src/m4

- *you may change the path in test.py*

- I recommand you to train model with Juna2/grasp_detection(which is also modified version of robot-grasp-detection) first and copy the model to the path.





## How to use

1. $ catkin_make

2.  
    $ roslaunch ros_grasp_detection

3. It subscribes "/croppedRoI" message which is 224x224 object image and publishes "/objects" which contains center x, y and degree. So you have to use another package to publish 224x224 object image to this package.

4. The grasp detection result will be saved in      

    /home/<your_path>/catkin_ws/src/ros_grasp_detection/src/image/bbox.jpg


- **One Thing you should know is that this test.py uses the model trained by modified robot-grasp-detection which is Juna2/grasp_detection.**



## sample

<img src=./src/image/example/bbox0.jpg width="30%"> <img src=./src/image/example/bbox1.jpg width="30%">
<br />
<img src=./src/image/example/bbox2.jpg width="30%"> <img src=./src/image/example/bbox3.jpg width="30%">



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package ros_grasp_detection
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.9.0 (2018-04-11)
------------------

0.8.1 (2017-07-27)
------------------

0.8.0 (2017-03-10)
------------------

0.7.1 (2016-10-24)
------------------
* add example of periodical publishing with rospy.Timer (`#34 <https://github.com/ros/ros_tutorials/issues/34>`_)

0.7.0 (2016-03-18)
------------------
* fix minor misleading comment (`#27 <https://github.com/ros/ros_tutorials/pull/27>`_)

0.6.1 (2015-09-19)
------------------

0.6.0 (2015-05-21)
------------------

0.5.3 (2015-05-04)
------------------

0.5.2 (2014-12-23)
------------------
* update code style of examples (`#18 <https://github.com/ros/ros_tutorials/pull/18>`_, `#19 <https://github.com/ros/ros_tutorials/pull/19>`_)

0.5.1 (2014-05-08)
------------------
* fix install when building Debian package

0.5.0 (2014-05-07)
------------------
* make rostest in CMakeLists optional (`ros/rosdistro#3010 <https://github.com/ros/rosdistro/issues/3010>`_)
* use queue_size for rospy.Publisher (`ros/ros_comm#169 <https://github.com/ros/ros_comm/issues/169>`_)
* use catkin_install_python() to install Python scripts (`#17 <https://github.com/ros/ros_tutorials/issues/17>`_)

0.4.3 (2014-01-07)
------------------
* python 3 compatibility (`#13 <https://github.com/ros/ros_tutorials/issues/13>`_)
* remove load_manifest calls

0.4.2 (2013-10-04)
------------------

0.4.1 (2013-09-11)
------------------
* check for CATKIN_ENABLE_TESTING

0.4.0 (2013-09-06)
------------------

0.3.13 (2013-08-21)
-------------------
* check for CATKIN_ENABLE_TESTING
* remove mainpage.dox

0.3.12 (2013-03-29)
-------------------

0.3.11 (2013-03-21)
-------------------
* update email in package.xml

0.3.10 (2013-03-08)
-------------------

0.3.9 (2012-12-21)
------------------
* modified dep type of catkin

0.3.8 (2012-12-13)
------------------
* add missing downstream depend
* switched from langs to message_* packages

0.3.7 (2012-12-06)
------------------
* fix node names (fix `#2 <https://github.com/ros/ros_tutorials/issues/2>`_)

0.3.6 (2012-10-30)
------------------
* fix catkin function order

0.3.5 (2012-10-18)
------------------

0.3.4 (2012-10-06)
------------------

0.3.3 (2012-10-05)
------------------
* added missing dep to rostest
* fixed missing genmsg stuff
* updated to latest catkin
* updated package.xml files
* added package.xml files

0.3.2 (2012-09-05)
------------------
* updated catkin variables

0.3.1 (2012-09-03)
------------------
* use install destination variables, removed manual installation of manifests

0.3.0 (2012-08-29)
------------------
* updated to current catkin

0.2.20 (2013-02-08)
-------------------

0.2.19 (2012-06-15 03:13:40 +0000)
----------------------------------
* make find_package REQUIRED
* removed obsolete catkin tag from manifest files
* fixed permissions of installed ros_grasp_detection files
* remove old Makefiles and bump to 0.2.13
* removed catkin_export_python() from ros_grasp_detection
* clean up build; tests turned on (but failing)
* adding <catkin/>, removing depends and platform tags
* updated to new catkin_export_python macro
* Don't install version control files
* catkin tweaks
* catkin tweaks
* initial add of ros_grasp_detection
* url fixes
* now uses myargv
* Added Ubuntu platform tags
* fixed deprecated use
* a2i_client: simplified conversion
* added msg/srv export
* Remove use of deprecated rosbuild macros
* updated to load_manifest
* Create trunk/branches/tags


src/test.py
#!/usr/bin/env python
import sys
import rospy
import numpy as np
import tensorflow as tf
import grasp_img_proc
from grasp_inf import inference
# from grasp_det import grasp_to_bbox
import cv2
from cv_bridge import CvBridge, CvBridgeError 
from ros_grasp_detection.msg import positionNpose
from sensor_msgs.msg import Image
import math
import sys 


class detector: 
    def __init__(self):
        # self.image_final = np.array(1, 224, 224, 3)
        self.bridge = CvBridge()
        self.count = 0

        self.X = tf.placeholder(tf.float32, [1,224,224,3], name="X")
        self.Y = tf.unstack(inference(self.X), axis=1)

        dg={}
        lg = ['w1', 'b1', 'w2', 'b2', 'w3', 'b3', 'w4', 'b4', 'w5', 'b5', 'w_fc1', 'b_fc1', 'w_fc2', 'b_fc2', 'w_output', 'b_output']
        # lg = ['w1', 'b1', 'w2', 'b2', 'w3', 'b3', 'w4', 'b4', 'w5', 'b5', 'w_fc1', 'b_fc1', 'w_fc2', 'b_fc2', 'w_fc3', 'b_fc3', 'w_output', 'b_output']
        for i in lg:
            dg[i] = [v for v in tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES) if v.name == i+':0'][0]
        self.saver_g = tf.train.Saver(dg)

        self._session = tf.InteractiveSession()
        
        self.saver_g.restore(self._session, '/home/juna/catkin_ws/src/ros_grasp_detection/src/models/grasp/m4/m4.ckpt')
        # init_op = tf.group(tf.global_variables_initializer(), tf.local_variables_initializer()) # init_op = tf.group(tf.global_variables_initializer(), tf.local_variables_initializer())
        # self._session.run(init_op)

        self.object_pub = rospy.Publisher("/objects", positionNpose, queue_size=1)
        self.image_sub = rospy.Subscriber("/croppedRoI", Image, self.image_callback, queue_size=1, buff_size=2**24)

    def draw_bbox(self, img, bbox):
        p1 = (int(float(bbox[0][0])), int(float(bbox[0][1])))
        p2 = (int(float(bbox[1][0])), int(float(bbox[1][1])))
        p3 = (int(float(bbox[2][0])), int(float(bbox[2][1])))
        p4 = (int(float(bbox[3][0])), int(float(bbox[3][1])))
        cv2.line(img, p1, p2, (255, 0, 0))
        cv2.line(img, p1, p3, (0, 0, 255))
        cv2.line(img, p4, p3, (0, 0, 255))
        cv2.line(img, p4, p2, (0, 0, 255))

        return img

    def grasp_to_bbox(self, x, y, x_vec, y_vec, h):
        hypo = math.sqrt(x_vec**2+y_vec**2)
        center = np.array([[1, 0, 0, x],
                           [0, 1, 0, y],
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]])
        rotate = np.array([[x_vec/hypo, -y_vec/hypo,   0,   0],
                           [y_vec/hypo,  x_vec/hypo,   0,   0],
                           [         0,           0,   1,   0],
                           [         0,           0,   0,   1]])
        trans_1 = np.array([[1, 0, 0,  h/2],
                            [0, 1, 0,    0],
                            [0, 0, 1,    0],
                            [0, 0, 0,    1]])
        trans_2 = np.array([[1, 0, 0, -h/2],
                            [0, 1, 0,    0],
                            [0, 0, 1,    0],
                            [0, 0, 0,    1]])
        trans_3 = np.array([[1, 0, 0,   0],
                            [0, 1, 0,  20],
                            [0, 0, 1,   0],
                            [0, 0, 0,   1]])
        trans_4 = np.array([[1, 0, 0,   0],
                            [0, 1, 0, -20],
                            [0, 0, 1,   0],
                            [0, 0, 0,   1]])
        point1 = [center.dot(rotate).dot(trans_1).dot(trans_3)[0,3], center.dot(rotate).dot(trans_1).dot(trans_3)[1,3]]
        point2 = [center.dot(rotate).dot(trans_1).dot(trans_4)[0,3], center.dot(rotate).dot(trans_1).dot(trans_4)[1,3]]
        point3 = [center.dot(rotate).dot(trans_2).dot(trans_3)[0,3], center.dot(rotate).dot(trans_2).dot(trans_3)[1,3]]
        point4 = [center.dot(rotate).dot(trans_2).dot(trans_4)[0,3], center.dot(rotate).dot(trans_2).dot(trans_4)[1,3]]

        
    # def grasp_to_bbox(self, x, y, tan, h, w):
    #     theta = math.atan(tan)
    #     edge1 = (x -w/2*math.cos(theta) +h/2*math.sin(theta), y -w/2*math.sin(theta) -h/2*math.cos(theta))
    #     edge2 = (x +w/2*math.cos(theta) +h/2*math.sin(theta), y +w/2*math.sin(theta) -h/2*math.cos(theta))
    #     edge3 = (x +w/2*math.cos(theta) -h/2*math.sin(theta), y +w/2*math.sin(theta) +h/2*math.cos(theta))
    #     edge4 = (x -w/2*math.cos(theta) -h/2*math.sin(theta), y -w/2*math.sin(theta) +h/2*math.cos(theta))      
      
        # edge1 = (x -w/2*math.sin(theta) -h/2*math.cos(theta), y -w/2*math.cos(theta) +h/2*math.sin(theta))
        # edge2 = (x +w/2*math.sin(theta) -h/2*math.cos(theta), y +w/2*math.cos(theta) +h/2*math.sin(theta))
        # edge3 = (x +w/2*math.sin(theta) +h/2*math.cos(theta), y +w/2*math.cos(theta) -h/2*math.sin(theta))
        # edge4 = (x -w/2*math.sin(theta) +h/2*math.cos(theta), y -w/2*math.cos(theta) -h/2*math.sin(theta))


        return [point1, point2, point3, point4]


    def image_callback(self, data):

        try:
            cv_image = self.bridge.imgmsg_to_cv2(data, "bgr8")
        except CvBridgeError as e:
            print(e, '!')
        img_resize = cv2.resize(cv_image, dsize=(224, 224), interpolation=cv2.INTER_CUBIC)
        
        image_np = np.array([np.asarray(img_resize)])
        # print(image_np)
        image_np = image_np * 1.0 / 255
        image_np = (image_np - 0.5) * 2

        '''There is some images which have 4 as a 3rd vector size like, 
        shape = (224, 224, 4) so you have to slice it so that they become like shape = (224, 224, 3)'''
        # img_slice = tf.slice(img_resize, [0,0,0], [224,224,3]) #### 
        # # print('2 : ', sess.run(img_reshape).shape)
        # img_reshape = tf.reshape(img_slice, shape=[1, 224, 224, 3])
        # cv2.imshow('hahaha', self._session.run(img_reshape)) 
        # cv2.waitKey(0) 
        # print('3 : ', img_reshape.get_shape())
        # x_hat, y_hat, tan_hat, w_hat, h_hat = tf.unstack(inference(img_reshape), axis=1)

        
        predict = self._session.run(self.Y, feed_dict={self.X:image_np})
        print 'x= ', predict[0], '\n y= ', predict[1], '\n x_vec= ', predict[2], '\n y_vec= ', predict[3], '\n h= ', predict[4]
        bbox_hat = self.grasp_to_bbox(predict[0], predict[1], predict[2], predict[3], predict[4]) # (math.atan(predict[2]))*180/3.14
        

        # n = bbox_hat[0][0].eval()
        # print(predict[0])
        img_final = self.draw_bbox(img_resize, bbox_hat)

        cv2.imwrite('/home/juna/catkin_ws/src/ros_grasp_detection/src/image/bbox'+str(self.count)+'.jpg', img_final)
        self.count += 1
        # cv2.waitKey(0)

        self.object_pub.publish(positionNpose(None, predict[0], predict[1], predict[2]))


def main(args):
    rospy.init_node('test')
    # print('4')
    obj=detector()

    try:
        rospy.spin()
    except KeyboardInterrupt:
        print("ShutDown")
        cv2.destroyAllWindows()

if __name__=='__main__':
    # print('3')
    main(sys.argv)
    # sess.close()




src/inference.py
'''
Inference model for grasping
'''
import tensorflow as tf
FLAGS = tf.app.flags.FLAGS

tf.app.flags.DEFINE_bool('trainable', False,
                            """Computes or not gradients for learning.""")

def conv2d_s2(x, W):
    return tf.nn.conv2d(x, W, strides=[1,2,2,1], padding='VALID') # padding='SAME'

def conv2d_s1(x, W):
    return tf.nn.conv2d(x, W, strides=[1,1,1,1], padding='VALID') # padding='SAME'
    
def max_pool_2x2(x):
    return tf.nn.max_pool(x, ksize=[1,2,2,1], strides=[1,2,2,1], padding='VALID') # padding='SAME'

def inference(images):
    if FLAGS.trainable:
        keep_prob = 0.8
    else:
        keep_prob = 1.
    print('keep_prob = %.1f' %keep_prob)

    w1 = tf.get_variable('w1', shape=[5,5,3,64], trainable=FLAGS.trainable)
    b1 = tf.get_variable('b1', initializer=tf.constant(0.1, shape=[64]), trainable=FLAGS.trainable)
    h1 = tf.nn.relu(conv2d_s2(images, w1)+b1)
    h1_pool = max_pool_2x2(h1)
    
    w2 = tf.get_variable('w2', [3,3,64,128], trainable=FLAGS.trainable)
    b2 = tf.get_variable('b2', initializer=tf.constant(0.1, shape=[128]), trainable=FLAGS.trainable)
    h2 = tf.nn.relu(conv2d_s2(h1_pool,w2)+b2)
    h2_pool = max_pool_2x2(h2)

    w3 = tf.get_variable('w3', [3,3,128,128], trainable=FLAGS.trainable)
    b3 = tf.get_variable('b3', initializer=tf.constant(0.1, shape=[128]), trainable=FLAGS.trainable)
    h3 = tf.nn.relu(conv2d_s1(h2_pool,w3)+b3)
    
    w4 = tf.get_variable('w4', [3,3,128,128], trainable=FLAGS.trainable)
    b4 = tf.get_variable('b4', initializer=tf.constant(0.1, shape=[128]), trainable=FLAGS.trainable)
    h4 = tf.nn.relu(conv2d_s1(h3,w4)+b4)

    w5 = tf.get_variable('w5', [3,3,128,256], trainable=FLAGS.trainable)
    b5 = tf.get_variable('b5', initializer=tf.constant(0.1, shape=[256]), trainable=FLAGS.trainable)
    h5 = tf.nn.relu(conv2d_s1(h4,w5)+b5)
    h5_pool = max_pool_2x2(h5)
    
    w_fc1 = tf.get_variable('w_fc1', [7*7*256,512], trainable=FLAGS.trainable)
    b_fc1 = tf.get_variable('b_fc1', initializer=tf.constant(0.1, shape=[512]), trainable=FLAGS.trainable)
    h5_flat = tf.reshape(h5_pool, [-1, 7*7*256])
    h_fc1 = tf.nn.relu(tf.matmul(h5_flat,w_fc1)+b_fc1)
    h_fc1_dropout = tf.nn.dropout(h_fc1, keep_prob)
    
    w_fc2 = tf.get_variable('w_fc2', [512,512], trainable=FLAGS.trainable)
    b_fc2 = tf.get_variable('b_fc2', initializer=tf.constant(0.1, shape=[512]), trainable=FLAGS.trainable)
    h_fc2 = tf.nn.relu(tf.matmul(h_fc1_dropout, w_fc2)+b_fc2)
    h_fc2_dropout = tf.nn.dropout(h_fc2, keep_prob)
    
    w_output = tf.get_variable('w_output', [512, 1000], trainable=FLAGS.trainable)
    b_output = tf.get_variable('b_output', [1000])
    output = tf.matmul(h_fc2_dropout,w_output)+b_output
    
    return output


src/build_cgd_dataset.py
#!/usr/bin/env python
'''Converts Cornell Grasping Dataset data into TFRecords data format using Example protos.
The raw data set resides in png and txt files located in the following structure:
    
    dataset/03/pcd0302r.png
    dataset/03/pcd0302cpos.txt
'''
'''
1. custom_dataset폴더를 만들고 그 안에 01, 02, 03, ..., 10 폴더를 만든다.
2. 그 폴더들 안에 각각 01은 pcd0101r.png ~ pcd0199r.png 
'''

import tensorflow as tf
import os
import glob
import numpy as np

dataset = '/home/juna/Documents/robot-grasp-detection/image/custom_dataset' # '/root/dataset/cornell_grasping_dataset''

class ImageCoder(object):
    def __init__(self):
        self._sess = tf.Session()
        self._decode_png_data = tf.placeholder(dtype=tf.string)
        self._decode_png = tf.image.decode_png(self._decode_png_data, channels=3)
    def decode_png(self, image_data):
        return self._sess.run(self._decode_png,
                              feed_dict={self._decode_png_data: image_data})

def _process_image(filename, coder):
    # Decode the image
    with open(filename) as f:
        image_data = f.read()
    image = coder.decode_png(image_data)
    assert len(image.shape) == 3
    height = image.shape[0]
    width = image.shape[1]
    assert image.shape[2] == 3    
    return image_data, height, width


def _process_bboxes(name):
    '''Create a list with the coordinates of the grasping rectangles. Every 
    element is either x or y of a vertex.'''
    with open(name, 'r') as f:
        bboxes = list(map(
              lambda coordinate: float(coordinate), f.read().strip().split()))
    return bboxes

def _int64_feature(v):
    if not isinstance(v, list):
        v = [v]
    return tf.train.Feature(int64_list=tf.train.Int64List(value=v))

def _floats_feature(v):
    if not isinstance(v, list):
        v = [v]
    return tf.train.Feature(float_list=tf.train.FloatList(value=v))

def _bytes_feature(v):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[v]))

def _convert_to_example(filename, bboxes, image_buffer, height, width):
    # Build an Example proto for an example
    example = tf.train.Example(features=tf.train.Features(feature={
          'image/filename': _bytes_feature(filename),
          'image/encoded': _bytes_feature(image_buffer),
          'image/height': _int64_feature(height),
          'image/width': _int64_feature(width),
          'bboxes': _floats_feature(bboxes)}))
    return example
    
def main():
    
    train_file = os.path.join(dataset, 'train-cgd')
    validation_file = os.path.join(dataset, 'validation-cgd')
    print(train_file)
    print(validation_file)
    writer_train = tf.python_io.TFRecordWriter(train_file)
    writer_validation = tf.python_io.TFRecordWriter(validation_file)
    
    # Creating a list with all the image paths
    folders = range(1,11)
    folders = ['0'+str(i) if i<10 else '10' for i in folders]
    filenames = []
    for i in folders:
        for name in glob.glob(os.path.join(dataset, i, 'pcd'+i+'*r.png')):
            filenames.append(name)
    
    # Shuffle the list of image paths
    np.random.shuffle(filenames)
    
    count = 0
    valid_img = 0
    train_img = 0

    coder = ImageCoder()
    for filename in filenames:
        bbox = filename[:-5]+'cpos.txt'
        bboxes = _process_bboxes(bbox)
        image_buffer, height, width = _process_image(filename, coder)
        example = _convert_to_example(filename, bboxes, image_buffer, height, width)
        # Split the dataset in 80% for training and 20% for validation
        # if count % 5 == 0:
        #     writer_validation.write(example.SerializeToString())
        #     valid_img +=1
        # else:
        writer_train.write(example.SerializeToString())
        train_img +=1
        count +=1
    
    print('Done converting %d images in TFRecords with %d train images and %d validation images' % (count, train_img, valid_img))
    
    writer_train.close()
    writer_validation.close()


if __name__ == '__main__':
    main()


src/grasp_det.py
#!/usr/bin/env python
''' 
Training a network on cornell grasping dataset for detecting grasping positions.
'''
import sys
import argparse
import os.path
import glob
import tensorflow as tf
import numpy as np
from shapely.geometry import Polygon
import grasp_img_proc
from grasp_inf import inference
import math
import time

'''This script can train and test image data by using option "--train_or_validation" but the test results are printed in terminal'''

TRAIN_FILE = '/home/juna/Documents/robot-grasp-detection/image/custom_dataset/train-cgd'  # '/root/dataset/cornell_grasping_dataset/train-cgd'
VALIDATE_FILE = '/home/juna/Documents/robot-grasp-detection/image/custom_dataset/validation-cgd'  # '/root/dataset/cornell_grasping_dataset/validation-cgd'

def bboxes_to_grasps(bboxes):
    # converting and scaling bounding boxes into grasps, g = {x, y, tan, h, w}
    box = tf.unstack(bboxes, axis=1)
    x = (box[0] + (box[4] - box[0])/2)
    y = (box[1] + (box[5] - box[1])/2)
    tan = (box[3] -box[1]) / (box[2] -box[0])
    h = tf.sqrt(tf.pow((box[2] -box[0]), 2) + tf.pow((box[3] -box[1]), 2))
    w = tf.sqrt(tf.pow((box[6] -box[0]), 2) + tf.pow((box[7] -box[1]), 2))
    return x, y, tan, h, w

def grasp_to_bbox(x, y, tan, h, w):
    theta = tf.atan(tan)
    # edge1 = (x -w/2*tf.cos(theta) +h/2*tf.sin(theta), y -w/2*tf.sin(theta) -h/2*tf.cos(theta))
    # edge2 = (x +w/2*tf.cos(theta) +h/2*tf.sin(theta), y +w/2*tf.sin(theta) -h/2*tf.cos(theta))
    # edge3 = (x +w/2*tf.cos(theta) -h/2*tf.sin(theta), y +w/2*tf.sin(theta) +h/2*tf.cos(theta))
    # edge4 = (x -w/2*tf.cos(theta) -h/2*tf.sin(theta), y -w/2*tf.sin(theta) +h/2*tf.cos(theta))

    edge1 = (x -w/2*math.sin(theta) -h/2*math.cos(theta), y -w/2*math.cos(theta) +h/2*math.sin(theta))
    edge2 = (x +w/2*math.sin(theta) -h/2*math.cos(theta), y +w/2*math.cos(theta) +h/2*math.sin(theta))
    edge3 = (x +w/2*math.sin(theta) +h/2*math.cos(theta), y +w/2*math.cos(theta) -h/2*math.sin(theta))
    edge4 = (x -w/2*math.sin(theta) +h/2*math.cos(theta), y -w/2*math.cos(theta) -h/2*math.sin(theta))
    return [edge1, edge2, edge3, edge4]

def run_training():
    print(FLAGS.train_or_validation)
    if FLAGS.train_or_validation == 'train':
        print('distorted_inputs')
        data_files_ = TRAIN_FILE
        images, bboxes = grasp_img_proc.distorted_inputs(
                  [data_files_], FLAGS.num_epochs, batch_size=FLAGS.batch_size)
    else:
        print('inputs')
        data_files_ = VALIDATE_FILE
        images, bboxes = grasp_img_proc.inputs([data_files_])
    
    x, y, tan, h, w = bboxes_to_grasps(bboxes)
    x_hat, y_hat, tan_hat, h_hat, w_hat = tf.unstack(inference(images), axis=1) # list
    # tangent of 85 degree is 11 
    tan_hat_confined = tf.minimum(11., tf.maximum(-11., tan_hat))
    tan_confined = tf.minimum(11., tf.maximum(-11., tan))
    # Loss function
    gamma = tf.constant(10.)
    loss = tf.reduce_sum(tf.pow(x_hat -x, 2) +tf.pow(y_hat -y, 2) + gamma*tf.pow(tan_hat_confined - tan_confined, 2) +tf.pow(h_hat -h, 2) +tf.pow(w_hat -w, 2))
    train_op = tf.train.AdamOptimizer(epsilon=0.1).minimize(loss)
    init_op = tf.group(tf.global_variables_initializer(), tf.local_variables_initializer())
    sess = tf.Session()
    sess.run(init_op)
    coord = tf.train.Coordinator()
    threads = tf.train.start_queue_runners(sess=sess, coord=coord)
    #save/restore model
    d={}
    l = ['w1', 'b1', 'w2', 'b2', 'w3', 'b3', 'w4', 'b4', 'w5', 'b5', 'w_fc1', 'b_fc1', 'w_fc2', 'b_fc2']
    for i in l:
        d[i] = [v for v in tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES) if v.name == i+':0'][0]
    
    dg={}
    lg = ['w1', 'b1', 'w2', 'b2', 'w3', 'b3', 'w4', 'b4', 'w5', 'b5', 'w_fc1', 'b_fc1', 'w_fc2', 'b_fc2', 'w_output', 'b_output']
    for i in lg:
        dg[i] = [v for v in tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES) if v.name == i+':0'][0]

    saver = tf.train.Saver(d)
    saver_g = tf.train.Saver(dg)
    #saver.restore(sess, "/root/grasp/grasp-detection/models/imagenet/m2/m2.ckpt")
    saver_g.restore(sess, FLAGS.model_path)
    try:
        count = 0
        step = 0
        start_time = time.time()
        while not coord.should_stop():
            start_batch = time.time()
            #train
            if FLAGS.train_or_validation == 'train':
                _, loss_value, x_value, x_model, tan_value, tan_model, h_value, h_model, w_value, w_model = sess.run([train_op, loss, x, x_hat, tan, tan_hat, h, h_hat, w, w_hat])
                duration = time.time() - start_batch
                if step % 100 == 0:             
                    print('Step %d | loss = %s\n | x = %s\n | x_hat = %s\n | tan = %s\n | tan_hat = %s\n | h = %s\n | h_hat = %s\n | w = %s\n | w_hat = %s\n | (%.3f sec/batch\n')%(step, loss_value, x_value[:3], x_model[:3], tan_value[:3], tan_model[:3], h_value[:3], h_model[:3], w_value[:3], w_model[:3], duration)
                if step % 1000 == 0:
                    if step == 1000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 2000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 3000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 4000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 5000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 6000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 7000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 8000:
                        saver_g.save(sess, FLAGS.model_path)
                    elif step == 10000:
                        coord.request_stop()
                    
            else:
                bbox_hat = grasp_to_bbox(x_hat, y_hat, tan_hat, h_hat, w_hat)
                bbox_value, bbox_model, tan_value, tan_model = sess.run([bboxes, bbox_hat, tan, tan_hat])
                bbox_value = np.reshape(bbox_value, -1)
                bbox_value = [(bbox_value[0],bbox_value[1]),(bbox_value[2],bbox_value[3]),(bbox_value[4],bbox_value[5]),(bbox_value[6],bbox_value[7])] 
                p1 = Polygon(bbox_value)
                p2 = Polygon(bbox_model)
                iou = p1.intersection(p2).area / (p1.area +p2.area -p1.intersection(p2).area) 
                angle_diff = np.abs(np.arctan(tan_model)*180/np.pi -np.arctan(tan_value)*180/np.pi)
                duration = time.time() -start_batch
                if angle_diff < 30. and iou >= 0.25:
                    count+=1
                    print('image: %d | duration = %.2f | count = %d | iou = %.2f | angle_difference = %.2f' %(step, duration, count, iou, angle_diff))
            step +=1
    except tf.errors.OutOfRangeError:
        print('Done training for %d epochs, %d steps, %.1f min.' % (FLAGS.num_epochs, step, (time.time()-start_time)/60))
    finally:
        coord.request_stop()

    coord.join(threads)
    sess.close()

def main(_):
    run_training()
    
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--learning_rate',
        type=float,
        default=0.001,
        help='Initial learning rate.'
    )
    parser.add_argument(
        '--data_dir',
        type=str,
        default='/root/imagenet-data',
        help='Directory with training data.'
    )
    parser.add_argument(
        '--num_epochs',
        type=int,
        default=None,
        help='Number of epochs to run trainer.'
    )
    parser.add_argument(
        '--batch_size',
        type=int,
        default=64,
        help='Batch size.'
    )
    parser.add_argument(
        '--log_dir',
        type=str,
        default='/tmp/tf',
        help='Tensorboard log_dir.'
    )
    parser.add_argument(
        '--model_path',
        type=str,
        default='/tmp/tf/model.ckpt',
        help='Variables for the model.'
    )
    parser.add_argument(
        '--train_or_validation',
        type=str,
        default='validation',
        help='Train or evaluate the dataset'
    )
    FLAGS, unparsed = parser.parse_known_args()
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
    

'''./grasp_det.py --model_path /home/juna/Documents/robot-grasp-detection/models/grasp/m4/m4.ckpt --train_or_validation validation'''
'''--num_epochs 5'''